From 5cd3f37c1a5957e0ad96bb763bb19edcab0467dd Mon Sep 17 00:00:00 2001
From: Nitin Garg <nitin.garg@freescale.com>
Date: Thu, 10 Jan 2013 14:14:58 -0600
Subject: [PATCH 146/310] ENGR00239687: Add emulated Storage support for iMX53
 SMD

Add emulated Storage support for iMX53 SMD as needed for
JB 4.2 multiuser experience.

Signed-off-by: Nitin Garg <nitin.garg@freescale.com>
---
 imx53_smd/init.freescale.rc                        | 24 +++++++++++++++++-----
 .../base/core/res/res/xml/storage_list.xml         |  4 ++--
 imx53_smd/vold.fstab                               |  2 +-
 imx5x/init.rc                                      | 15 ++++++++------
 4 files changed, 31 insertions(+), 14 deletions(-)

diff --git a/imx53_smd/init.freescale.rc b/imx53_smd/init.freescale.rc
index 32fbb76..c9e62e1 100755
--- a/imx53_smd/init.freescale.rc
+++ b/imx53_smd/init.freescale.rc
@@ -3,6 +3,20 @@ import init.${ro.hardware}.usb.rc
 on init
     start watchdogd
 
+    # See storage config details at http://source.android.com/tech/storage/
+    mkdir /mnt/shell/emulated 0700 shell shell
+    mkdir /storage/emulated 0555 root root
+
+    export EXTERNAL_STORAGE /storage/emulated/legacy
+    export EMULATED_STORAGE_SOURCE /mnt/shell/emulated
+    export EMULATED_STORAGE_TARGET /storage/emulated
+
+    # Support legacy paths
+    symlink /storage/emulated/legacy /sdcard
+    symlink /storage/emulated/legacy /mnt/sdcard
+    symlink /storage/emulated/legacy /storage/sdcard0
+    symlink /mnt/shell/emulated/0 /storage/emulated/legacy
+
 on boot
 
     # Set permission for IIM node
@@ -69,11 +83,11 @@ on boot
     mount debugfs none /sys/kernel/debug/
 
 # Daemon processes to be run by init.
-service hciattach /system/bin/logwrapper /system/bin/hciattach -n -s 115200 /dev/ttymxc2 ath3k 3000000 flow sleep
-    user root
-    group bluetooth net_bt_admin
-    disabled
-    oneshot
+#service hciattach /system/bin/logwrapper /system/bin/hciattach -n -s 115200 /dev/ttymxc2 ath3k 3000000 flow sleep
+#    user root
+#    group bluetooth net_bt_admin
+#    disabled
+#    oneshot
 
 service p2p_supplicant /system/bin/wpa_supplicant \
     -ip2p0 -Dnl80211 -c/data/misc/wifi/p2p_supplicant.conf -N \
diff --git a/imx53_smd/overlay/frameworks/base/core/res/res/xml/storage_list.xml b/imx53_smd/overlay/frameworks/base/core/res/res/xml/storage_list.xml
index f486f92..84cce20 100755
--- a/imx53_smd/overlay/frameworks/base/core/res/res/xml/storage_list.xml
+++ b/imx53_smd/overlay/frameworks/base/core/res/res/xml/storage_list.xml
@@ -33,9 +33,9 @@
 -->
 
 <StorageList xmlns:android="http://schemas.android.com/apk/res/android">
-    <storage android:mountPoint="/mnt/sdcard"
-             android:storageDescription="@string/storage_internal"
+    <storage android:storageDescription="@string/storage_internal"
              android:primary="true"
+             android:emulated="true"
              android:mtpReserve="100" />
     <storage android:mountPoint="/mnt/extsd"
              android:storageDescription="@string/storage_sd_card"
diff --git a/imx53_smd/vold.fstab b/imx53_smd/vold.fstab
index 0f779f6..561fb87 100644
--- a/imx53_smd/vold.fstab
+++ b/imx53_smd/vold.fstab
@@ -15,7 +15,7 @@
 ## Example of a standard sdcard mount for the emulator / Dream
 # Mounts the first usable partition of the specified device
 #for internal emmc as primary storage
-dev_mount sdcard /mnt/sdcard 1 /devices/platform/sdhci-esdhc-imx.0/mmc_host/mmc0 /devices/platform/sdhci-esdhc-imx.2/mmc_host/mmc0
+#dev_mount sdcard /mnt/sdcard 1 /devices/platform/sdhci-esdhc-imx.0/mmc_host/mmc0 /devices/platform/sdhci-esdhc-imx.2/mmc_host/mmc0
 #for external sdcard as primary storage
 dev_mount extsd /mnt/extsd auto /devices/platform/sdhci-esdhc-imx.2/mmc_host/mmc2 /devices/platform/sdhci-esdhc-imx.0/mmc_host/mmc2
 #used for usb host
diff --git a/imx5x/init.rc b/imx5x/init.rc
index ef880b0..b329eae 100644
--- a/imx5x/init.rc
+++ b/imx5x/init.rc
@@ -34,12 +34,11 @@ loglevel 3
     export ANDROID_ROOT /system
     export ANDROID_ASSETS /system/app
     export ANDROID_DATA /data
+    export ANDROID_STORAGE /storage
     export ASEC_MOUNTPOINT /mnt/asec
     export LOOP_MOUNTPOINT /mnt/obb
-    export SHM_MOUNTPOINT /mnt/shm
-    export EXTERNAL_STORAGE /mnt/sdcard
     export BOOTCLASSPATH /system/framework/core.jar:/system/framework/core-junit.jar:/system/framework/bouncycastle.jar:/system/framework/ext.jar:/system/framework/framework.jar:/system/framework/telephony-common.jar:/system/framework/mms-common.jar:/system/framework/android.policy.jar:/system/framework/services.jar:/system/framework/apache-xml.jar
-
+    export SHM_MOUNTPOINT /mnt/shm
     export LIGHTSENSOR_PATH /sys/class/i2c-adapter/i2c-1/1-0044
     export VPU_FW_PATH /system/lib/firmware/vpu
     export CODEC_SHM_PATH /mnt/shm
@@ -62,8 +61,6 @@ loglevel 3
     mount cgroup none /acct cpuacct
     mkdir /acct/uid
 
-    mkdir /mnt/sdcard 0000 system system
-    symlink /mnt/sdcard /sdcard
     mkdir /mnt/extsd 0000 system system
     symlink /mnt/extsd /extsd
     mkdir /mnt/udisk 0000 system system
@@ -205,6 +202,7 @@ on post-fs-data
     write /proc/apanic_console 1
 
     mkdir /data/media 0770 media_rw media_rw
+    mount vfat /dev/block/mmcblk0p1 /data/media fmask=0007,dmask=0007,allow_utime=0020,uid=1023,gid=1023
 
     # create basic filesystem structure
     mkdir /data/misc 01771 system misc
@@ -282,7 +280,8 @@ on boot
     chmod 0664 /sys/module/lowmemorykiller/parameters/minfree
 
     # Tweak background writeout
-    write /proc/sys/vm/dirty_expire_centisecs 200
+    write /proc/sys/vm/dirty_expire_centisecs 50
+    write /proc/sys/vm/dirty_writeback_centisecs 100
     write /proc/sys/vm/dirty_background_ratio  5
 
     # Permissions for System Server and daemons.
@@ -625,6 +624,10 @@ service bugreport /system/bin/bugmailer.sh -b /system/media/audio/ui/Dock.ogg -e
     oneshot
     keycodes 115 102
 
+# virtual sdcard daemon running as media_rw (1023)
+service sdcard /system/bin/sdcard /data/media /mnt/shell/emulated 1023 1023
+    class late_start
+
 # Set watchdog timer to 30 seconds and pet it every 10 seconds to get a 20 second margin
 service watchdogd /sbin/watchdogd 10 20
     class core
-- 
1.8.0

